registry/client/auth/authchallenge.go

```diff
@@ -54,12 +65,65 @@ func init() {
 	}
 }
 
-func parseAuthHeader(header http.Header) map[string]authorizationChallenge {
-	challenges := map[string]authorizationChallenge{}
+type ErrAuthorizationChallenge struct {
+	Challenges []Challenge
+}
+
+func (err *ErrAuthorizationChallenge) Error() string {
+	schemes := make([]string, len(err.Challenges))
+	for i, challenge := range err.Challenges {
+		schemes[i] = challenge.Scheme
+	}
+	return fmt.Sprintf("authorization required using %s", strings.Join(schemes, " or "))
+}
+
+// Check gets the API versions out of an HTTP response using the provided
+// version header as the key for the HTTP header. If the http response was
+// unauthorized an `ErrAuthorizationChallenge` is thrown with the provided
+// Challenges. Any other response besides an OK is considered invalid.
+func Check(resp *http.Response, versionHeader string) ([]APIVersion, error) {
+	versions := []APIVersion{}
+	if versionHeader != "" {
+		for _, supportedVersions := range resp.Header[http.CanonicalHeaderKey(versionHeader)] {
+			for _, version := range strings.Fields(supportedVersions) {
+				versions = append(versions, parseVersion(version))
+			}
+		}
+	}
+
+	if resp.StatusCode == http.StatusUnauthorized {
+		// Parse the WWW-Authenticate Header and store the challenges
+		// on this endpoint object.
+		return versions, &ErrAuthorizationChallenge{
+			Challenges: parseAuthHeader(resp.Header),
+		}
+	} else if resp.StatusCode != http.StatusOK {
+		return versions, fmt.Errorf("invalid response: %d", resp.StatusCode)
+	}
+
+	return versions, nil
+}
+
+func parseVersion(versionStr string) APIVersion {
```

Comment:
This function should be under the `APIVersion` struct definition. Also, please document the function.